include/zephyr/sys/barrier_builtin.h

```diff
@@ -0,0 +1,31 @@
+/* barrier operations */
+
+/*
+ * Copyright (c) 2023 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_BARRIER_BUILTIN_H_
+#define ZEPHYR_INCLUDE_SYS_BARRIER_BUILTIN_H_
+
+#ifndef ZEPHYR_INCLUDE_SYS_BARRIER_H_
+#error Please include <zephyr/sys/barrier.h>
+#endif
+
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static ALWAYS_INLINE void z_barrier_dmem_fence_full(void)
+{
+	__atomic_thread_fence(__ATOMIC_SEQ_CST);
```

Comment:
This is a GCC built-in and will only work with GCC (and, at most, with Clang which intends to be compatible with GCC).

Since Zephyr kernel needs to be able to build with toolchains other than GCC, there should be an additional backend using the C11 `atomic_thread_fence`.

Reply:
but shouldn't this be a problem also for the atomic built-ins in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/atomic_builtin.h ?

Reply:
what about using
```c
#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
```
to switch to `atomic_thread_fence` in the same backend?

Reply:
Added https://github.com/zephyrproject-rtos/zephyr/pull/57439/commits/7da73c20668eb3ac77f0a36ec0178cfab675bc2e . Is this ok?

Reply:
Added review comment here: https://github.com/zephyrproject-rtos/zephyr/pull/57439#discussion_r1187394647